### DISCUSSION TITLE
Support separate mountpoint for metadata (split data on HDD and metadata on SSD)

### DISCUSSION THREAD

**Discussion by skrivy** [minio/minio#13972] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

Using rclone to create a copy to S3 generates for each object a metadata request. It overloads rotational drives as there are 
plenty of small objects.

**Describe the solution you'd like**

We would like to move metadata (12GB) to SSD in order to keep the load small and use cheap HDD storage for the rest (20TB).

**Describe alternatives you've considered**

We were looking at ZFS with it's cache, but it didn't fix our issue as most of the metadata weren't in the cache.

**Additional context**

Minio version:

```
minio version RELEASE.2021-12-10T23-03-39Z
```

When trying to move metadata to another drive, we get this error message:

```
ERROR Invalid command line arguments: Cross-device mounts detected on path (/var/www/minio) at following locations [/var/www/minio/.minio.sys/buckets]. Export path should not have any sub-mounts, refusing to start.
```

Aslo tested with whole `.minio.sys` directory:

```
ERROR Invalid command line arguments: Cross-device mounts detected on path (/var/www/minio) at following locations [/var/www/minio/.minio.sys]. Export path should not have any sub-mounts, refusing to start.
```

Related issue: [#10480](https://github.com/minio/minio/issues/10480)

**Comment by klauspost** [comment, marked as answer]
This is not a feature we will support an NAS/single drive mode.

For distributed setups you can use [tiering](https://docs.min.io/minio/baremetal/lifecycle-management/lifecycle-management-overview.html#object-transition-tiering) to achieve this.

**Reply by skrivy** [reply]
If I understand it right. 2 instances of minio should be created. One on SSDs the second one on HDDS. Right?
Thats quite complex to be honest.

**Comment by smiba** [comment]
Unfortunate to hear metadata is not going to be stored separately on a mountpoint, it would greatly reduce I/O on storage media where seeks are expensive (rotational media)

For anyone reading this, you can probably get something like this working by using a metadata special device in ZFS with special_small_blocks set to 64K or 32K. This means that small files (like the metadata by Minio) are gonna be stored on the special device (which has to be high IOPS storage like an SSD)

Unfortunately figured this out too late, so I'm now stuck with slow directory traversing

**Reply by harshavardhana** [reply]
you should think about the hot-tier -> warm-tier model @smiba

**Reply by skrivy** [reply]
As far as I understand it, minio tiering isn't much helpful in situations when you need crawl whole store - e.g. backup whole S3 to another S3. In this case, you will hit all metadata very hard and cause very high IO.

For comparasion - ceph supports multiple backends for metadata and data, so you can easily put metadata on ssd and have data on rotational drives.

**Reply by harshavardhana** [reply]
It is our prime directive not to separate metadata and data. We won't be making such changes - if Ceph gives you what you want you should use that @skrivy. Just let us know how things pan out at scale.

**Reply by smiba** [reply]
@harshavardhana Out of genuine intrest, how do other users deal with HDD backed storage with lots of objects? Surely there are others that would run into this issue where crawling is severely limited in speed due to this. It would be useless to store entire files on SSDs, especially considering HDD's are plenty fast in an array as long as they don't have to constantly seek.

Unfortunate to hear there are no plans on implementing this, I think it would be a worthy addition to Minio for both personal users and professional users. So many software supports similar (MariaDB with indexes on other volumes, Ceph with metadata, ZFS with metadata, and the list goes on)

**Reply by harshavardhana** [reply]
@smiba We have already designed a hot-tier -> warm-tier deployment model that perfectly fits this.  We have already deployed this at multiple petabytes at scale. With only 16 NVMe disks, 28Ti size can absorb 720 HDD worth of I/O at the hot tier. The number of objects going in is 5 billion or more.

Ceph's approach is incorrect. In any system, if there are many different ways with varying tolerance levels, SLAs invite trouble when key things like metadata are separate from data. 

In the real world, things break, and one should design a system to be simple and predictable under such scenarios. Having layers of different kinds of tolerance per information block will not scale, let alone debug in a production environment. Complexity inherently leads to failure, no matter how well you have engineered it. 

The software has to be simple to scale, easier to manage, and easier to debug in production. 

MinIO does not aim to build the perfect cockpit storage. 

I don't expect everyone to understand it, but that's okay 🍰

**Reply by smiba** [reply]
That's fair! Thanks for your explanation, makes sense! 

I guess if someone /really/ wants this they can always use the ZFS option I talked about earlier at their own risk. Should've just done so myself but unfortunately figured this out after the fact, learning moment :)

**Reply by FourSpaces** [reply]
Hi!  @harshavardhana  I see that as the number of objects in minio increases, obtaining the list of objects becomes very slow. I have an idea:

1. Establish a standalone KV database (boltdb) on each hard drive.

2. After performing operations on each object, update the metadata to a standalone kV database.

3. When obtaining the list, directly read the single machine kV data.


This way, metadata and data can be stored on a single hard drive without affecting the existing architecture.

Getting the object list will also be done quickly.

Modifying the code will only modify the parts related to metadata operations, and the rest will not be changed.

**Reply by smiba** [reply]
For what it's worth, I've been using ZFS with a "special matadata" device attached to my rotational disk array. (With the option enabled to store small files =<128KB on the special SSD)

This basically turned getting 1000 items using ListObjectsV2 from 15 seconds to 200ms

A ZFS special device as I described above seems to be the only solution as long as MinIO isn't providing any options to improve file listing speeds. Reading 1000 scattered single .xl files is just not doable on rotational disks

Would be cool if another solution did exist though like you described @FourSpaces, would love to see it!

**Reply by julienlau** [reply]
> Ceph's approach is incorrect. In any system, if there are many different ways with varying tolerance levels, SLAs invite trouble when key things like metadata are separate from data.
> 
> In the real world, things break, and one should design a system to be simple and predictable under such scenarios. Having layers of different kinds of tolerance per information block will not scale, let alone debug in a production environment. Complexity inherently leads to failure, no matter how well you have engineered it.
> 
> The software has to be simple to scale, easier to manage, and easier to debug in production.
> 
> MinIO does not aim to build the perfect cockpit storage.
> 
> I don't expect everyone to understand it, but that's okay 🍰

Thanks for clarifying this. This is very interesting, but I don't fully agree.
I find it really nice to have 2 solid open source implementations of object store with ceph and minio.
However I don't find that a 2 tier minio setup is inherently simpler than a ceph setup especially when dealing with on-premises platforms. Minio needs to have exactly the same hardware setup on all servers of a tier and uses a frozen topology once bootstraped. This is VERY hard to operate on-premises.
I would say the usual answer : it depends on the use case !

In my view as long as storage growth is non predictible and non linear, Ceph wins due to this problem of frozen infrastructure once started.

**Reply by skrivy** [reply]
> In the real world, things break, and one should design a system to be simple and predictable under such scenarios.

I believe sysadmin should have the option to decide whether he risks losing data or not. Once I require to split data and metadata I am pretty sure I know what I do and what consequences I can expect.

Tiering does not solve situations where you need to crawl whole storage and find changes (such as listing changes to backup).
